### PR TITLE
repo v2 support

### PR DESCRIPTION
**Pull request by bnewbold**
A couple high-level notes:

- reasonably confident in MST changes. not very confident in repo node changes (eg, `repo/` and `repomgr/` sections). I basically just updated the structs and got things to compile, the entire `repo` package could probably be simplified
- there is at least one call-out for closer review of some code I ripped out (in repomgr)
- I haven't verified that the signature process validates against atproto repo signatures (or vica versa)
- the MST changes also include new constraints on keys (paths): total length <= 256 bytes; must have a single '/'; small subset of ASCII chars allowed in keys
- I implemented the []byte / string mapping the same way as the typestring package, which means most of the logic is still string-based not bytes-based; conversation happens at the last moment when serializing/deserializing

### CODE REVIEW

File: repo/repo.go [hunk 1]
```diff
@@ -17,38 +19,57 @@ import (
...
+}
+
+// returns SHA-256 (as direct bytes, not hex-encoded string bytes) of the
+// DAG-CBOR representation of object. This is what gets signed.
+func (uc *UnsignedCommit) BytesForSigning() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	if err := uc.MarshalCBOR(buf); err != nil {
+		return []byte{}, err
+	}
+	h := sha256.Sum256(buf.Bytes())
```
Comment (whyrusleeping) [inline repo/repo.go:69]: I believe the signing library hashes the input already, need to verify

Comment (whyrusleeping) [reply repo/repo.go:69]: Confirmed (go-did/key.go.(PrivKey).Sign) hashes the input with sha256 before signing internally.
I'm open to doing that elsewhere, but it seems less error prone there

File: repomgr/repomgr.go [hunk 2]
```diff
@@ -957,33 +967,7 @@ func (rm *RepoManager) processNewRepo(ctx context.Context, user util.Uid, r io.R
...
-			return err
-		}
-
-		if err := membs.Put(ctx, blk); err != nil {
-			return err
-		}
-
-		membs = util.NewReadThroughBstore(robs, membs)
-	}
+	// TODO(bnewbold): @Why should review the code that was removed here
```
Comment (whyrusleeping) [inline repomgr/repomgr.go:970]: This is good, this was the hack necessary for the meta object being a separate thing. I'm only sad that i wasnt the one to delete it ;)

**Review by whyrusleeping — CHANGES_REQUESTED** [review]
This all looks good except for the hashing of the signable content

File: repo/repo.go (hunk 1 above)
Comment (bnewbold) [reply repo/repo.go:69]: Ok, I will pull this hash out.

File: mst/mst.go [hunk 3]
```diff
@@ -191,8 +191,9 @@ func (mst *MerkleSearchTree) getEntries(ctx context.Context) ([]NodeEntry, error
 func entriesFromNodeData(ctx context.Context, nd *NodeData, cst cbor.IpldStore) ([]NodeEntry, error) {
 	layer := -1
 	if len(nd.Entries) > 0 {
+		// NOTE(bnewbold): can compute the layer on the first KeySuffix, because for the first entry that field is a complete key
 		firstLeaf := nd.Entries[0]
-		layer = leadingZerosOnHash(firstLeaf.KeySuffix)
+		layer = leadingZerosOnHash(string(firstLeaf.KeySuffix))
```
[the remark this answers is not in the document]
Comment (bnewbold) [reply mst/mst.go:196]: Yup! As just discussed, keeping it as directly similar to the Typescript implementation for now, will refactor to be more idiomatic golang in probably not too long.

File: repomgr/repomgr.go (hunk 2 above)
Comment (bnewbold) [reply repomgr/repomgr.go:970]: Ok, I'll pull this comment out.

**Comment by bnewbold** [comment]
@whyrusleeping verification change LGTM, thanks for adding the test.

I'm ready for this to merge if you are.

**Review by whyrusleeping — APPROVED** [review]

**Pull request merged**